src/translate-virtual-address.h

```diff
@@ -187,39 +211,39 @@ static NO_INLINE bool translate_virtual_address(STATE_ACCESS &a, uint64_t *ppadd
         if (xwr != 0) {
             // These protection bit combinations are reserved for future use
             if (unlikely(xwr == PTE_XWR_W_MASK || xwr == (PTE_XWR_W_MASK | PTE_XWR_X_MASK))) {
-                return false;
+                break;
             }
-            // (We know we are not PRV_M if we reached here)
-            if (priv == PRV_S) {
-                if ((pte & PTE_U_MASK)) {
-                    // S-mode can never execute instructions from user pages, regardless of the state of SUM
-                    if (unlikely(xwr_shift == PTE_XWR_X_SHIFT)) {
-                        return false;
-                    }
-                    // If SUM is not set, forbid S-mode code from accessing U-mode memory
-                    if (unlikely(!(mstatus & MSTATUS_SUM_MASK))) {
-                        return false;
-                    }
+
+            // when checking the U bit for G translation, the current privilege mode is always taken to be U-mode
+            if constexpr (TRANSLATION_MODE == TRANSLATION_G) {
+                if (!(pte & PTE_U_MASK)) {
+                    break;
                 }
-            } else {
-                // Forbid U-mode code from accessing S-mode memory
-                if (unlikely(!(pte & PTE_U_MASK))) {
-                    return false;
+            }
+            if constexpr (TRANSLATION_MODE == TRANSLATION_VS) {
+                if (pte & PTE_U_MASK) {
```

Comment:
This code could benefit from a comment of what is going on here. Preferably quoting the spec if possible.